src/apps/lwaftr/fragmentv4.lua

```diff
@@ -0,0 +1,100 @@
+module(..., package.seeall)
+
+local constants = require("apps.lwaftr.constants")
+local packet = require("core.packet")
+local ffi = require("ffi")
+
+local band, bor, lshift = bit.band, bit.bor, bit.lshift
+local ntohs, htons = ffi.C.ntohs, ffi.C.htons
+
+local int16_ptr_t = ffi.typeof("uint16_t*")
+
+local total_length_offset = constants.ethernet_header_size + constants.o_ipv4_total_length
+local frag_id_offset = constants.ethernet_header_size + constants.o_ipv4_identification
+local flags_and_frag_offset_offset = constants.ethernet_header_size + constants.o_ipv4_flags
+local ihl_and_ver_offset = constants.ethernet_header_size + constants.o_ipv4_ver_and_ihl
+
+-- Constants to manipulate the flags next to the frag-offset field directly
+-- as a 16-bit integer, without needing to shift the 3 flag bits.
+local flag_dont_fragment_mask  = 0x4000
+local flag_more_fragments_mask = 0x2000
+local frag_offset_field_mask   = 0x1FFF
+
+
+local function get_u16_at_offset(pkt, offset)
+	return ntohs(ffi.cast(int16_ptr_t, pkt.data + offset)[0])
+end
+
+local function set_u16_at_offset(pkt, offset, value)
+	ffi.cast(int16_ptr_t, pkt.data + offset)[0] = htons(value)
+end
+
```

Comment:
Sure, it would be nice to have those in a module somewhere, and reuse them. Maybe even somewhere in Snabb. A potential issue is that even when inlined this kind of functions might not perform as fast as possible because of the call to `ffi.cast()` in each invocation, because if I am reading the LuaJIT code correctly, `ffi.cast()` pushes a new userdata to the Lua world each time. When doing lots of peek/poke operations, we would want to do the cast once, and then use indexes to the returned arrays (instead of offsets). Let met illustrate with an example (simplified):

```lua
local ver_and_ihl = get_u16_at_offset(pkt, ethernet_header_size)
local flags_frag_off = get_u16_at_offset(pkt, ethernet_header_size + o_ipv4_flags)
```

The above could be rewritten as:

```lua
local pkt_as_u16_array = ffi.cast(uint16_ptr_t, pkt.data + ethernet_header_size)
local ver_and_ihl = C.ntohs(pkt_as_u16_array[0])
local flags_frag_off = C.ntohs(pkt_as_u16_array[o_ipv4_flags / 2])
```

Now imagine the same but with a dozen of calls to peek/poke 16-bit values: saving the need to do the cast each time suddenly looks like a compelling idea. WDYT?